contributors/design-proposals/storage/volume-reconstruct.md

```diff
@@ -0,0 +1,118 @@
+* bind mount path (also pod volume dir): pods/podUID/volumes/kubernetes.io~gce-pd/my-volume
+
+So look more closely, you can see for only the purpose of volume tear down, if we have a record of those mount paths (bind and global) in actual state, there is no need to 
+construct those paths from the volume spec again. In case of kubelet restarts, those mount paths could be scanned from the mount lists left on the disk. (bind mount path has
+a path structure which can be easily idenetified and global mount path could be obtained from the mount reference of the bind mount path). This observation leads to our first
+possible solution.
+
+### Solution1: Remove the dependency of volume spec. 
+Modify the data structure in actual state to keep a record of the mount paths so that volume tear down process (unmount) can use those records instead of reconstructing them
+directly from volume spec. In case the kubelet restarts, reconstruction process just need to find out the left mounting paths and mark them into the actual state. Even if the
+volume spec cannot be reconstructed correctly, volume tear down process can still work. 
+
+Issue: Because some plugins cannot reconstruct volume spec correctly, which means the generated unique volume name might not match with the real one.
+In case another new pod is added to use the same volume, two volume operations on the same device might happen concurrently and cause race condition because they have different
+unique volume names.
+
+### Solution2: Cleanup mounts directly without reconstruction.
+This approach will try to clean up mounts that are no longer needed before reconciler starts to work. It scans the pod volume directories when kubelet restarts to find the mount
+paths (bind and global). If the pod volume no longer exist in desired state, clean up the mounts. To avoid long delay or mount hanging problem, we can use operation executor to
+start a new routine. 
+
+Issue: The potential issue of this approach is that cleanup process might fail and
+reconciler will not try to do this process again.
+
+## Proposal
+
+The current proposal is trying to take the advantages from the two possible
+solutions mentioned above. 
+1. If volume plugins can support reconstruction well (i.e., volume spec can
+   be fully recovered), reconciler will update this information into the states
+   and reconciler will take care of volume cleanup. Reconciler can retry if
+   cleanup fails.
+
+Currently, the following volume plugins can support reconstruction. These volume
+source names are just the device names which can be retrived from the mount paths.
+
+AWS EBS, Azure_dd, Cinder, CSI, GCE_PD, FC, iScsi, Photon, Rbd, Vsphere 
+
+2. If volume plugins cannot support reconstruction, reconciler clean up the
+   mounts directly with a go routine (operation executor) to avoid delay. If it
+   fails, reconciler will not retry.
+
+The following volume plugins cannot support reconstruction,
```

Comment:
Do we need to add `local volume plugin` to this group? @msau42 @dhirajh

Reply:
I think local volume can support reconstruction even though the spec is not complete.

For filesystem volumes: it only needs the kubelet's pod volume path
For block volumes: it doesn't do anything on teardown and just relies on common kubelet teardown

Do you see any other issues with an incomplete spec?  The incomplete spec means that Setup will fail, but we only care about Teardown here right?

Reply:
>> Do you see any other issues with an incomplete spec?

I think if the incomplete spec satisfies following conditions,  there is no other issue for reconstruction.
- for filesystem: plugin.NewMounter() should be succeeded
- for mapperPlugin.NewBlockVolumeMapper() should be succeeded and also spec should have `volumeMode`.

>> but we only care about Teardown here right?

Right.